File: backend/app/api/prompts.py
"""CRUD endpoints for PromptTemplate management"""
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.ext.asyncio import AsyncSession
from ..db.session import get_db
from ..api.deps import get_current_active_user
from ..schemas.prompt import PromptCreate, PromptUpdate, Prompt
from ..crud.prompt import prompt
from ..crud.favorite import favorite
from ..models.user import User
from ..models.prompt import PromptTemplate


router = APIRouter()


@router.post("/", response_model=Prompt)
async def create_prompt(
    prompt_in: PromptCreate,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
) -> PromptTemplate:
    """Create a new prompt template (creator set from current user)."""
    prompt_obj = await prompt.create(db, obj_in=prompt_in, creator=current_user)
    return prompt_obj


@router.get("/", response_model=List[Prompt])
async def read_prompts(
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db),
    skip: int = 0,
    limit: int = 100,
    q: str = Query(None, description="Full-text search query"),
    creator_id: int = Query(None, description="Filter by creator ID"),
    tags: str = Query(None, description="Comma-separated tags to filter by")
) -> List[PromptTemplate]:
    """List prompts with optional full-text search and filters (respects visibility)."""
    if q:
        prompts = await prompt.search(
            db, q, skip, limit, creator_id, tags, current_user.id
        )
    else:
        prompts = await prompt.get_multi(db, skip=skip, limit=limit, current_user_id=current_user.id)
    return prompts


@router.get("/{prompt_id}", response_model=Prompt)
async def read_prompt(
    prompt_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
) -> PromptTemplate:
    """Get a specific prompt by ID (respects visibility)."""
    prompt_obj = await prompt.get(db, prompt_id, current_user.id)
    if not prompt_obj:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Prompt not found or not accessible"
        )
    return prompt_obj


@router.put("/{prompt_id}", response_model=Prompt)
async def update_prompt(
    prompt_id: int,
    prompt_in: PromptUpdate,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
) -> PromptTemplate:
    """Update a prompt template (only creator can update)."""
    prompt_obj = await prompt.get(db, prompt_id, current_user.id)
    if not prompt_obj:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Prompt not found"
        )
    
    if prompt_obj.creator_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not enough permissions"
        )
    
    prompt_obj = await prompt.update(db, db_obj=prompt_obj, obj_in=prompt_in)
    return prompt_obj


@router.delete("/{prompt_id}")
async def delete_prompt(
    prompt_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
) -> dict:
    """Delete a prompt template (only creator can delete)."""
    prompt_obj = await prompt.get(db, prompt_id, current_user.id)
    if not prompt_obj:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Prompt not found"
        )
    
    if prompt_obj.creator_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not enough permissions"
        )
    
    await prompt.remove(db, id=prompt_id)
    return {"msg": "Prompt deleted"}


@router.post("/{prompt_id}/favorite")
async def add_favorite(
    prompt_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
) -> dict:
    """Add prompt to current user's favorites."""
    success = await favorite.add_favorite(db, user_id=current_user.id, prompt_id=prompt_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail="Prompt already in favorites"
        )
    return {"msg": "Prompt added to favorites"}


@router.delete("/{prompt_id}/favorite")
async def remove_favorite(
    prompt_id: int,
    current_user: User = Depends(get_current_active_user),
    db: AsyncSession = Depends(get_db)
) -> dict:
    """Remove prompt from current user's favorites."""
    success = await favorite.remove_favorite(db, user_id=current_user.id, prompt_id=prompt_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Favorite not found"
        )
    return {"msg": "Prompt removed from favorites"}


__all__ = ["router"]

File: backend/src/routes/prompts.js
const express = require('express');
const Joi = require('joi');
const { Op, where, literal, fn, col } = require('sequelize');
const { PromptTemplate, User, UserPromptFavorites } = require('../models/index.js');
const { authenticateToken } = require('../utils/auth.js');

const router = express.Router();

// Validation schemas
const promptCreateSchema = Joi.object({
  title: Joi.string().min(1).max(200).required(),
  short_description: Joi.string().min(10).max(500).required(),
  long_description: Joi.string().max(5000).optional(),
  template: Joi.string().min(10).required(),
  inputs: Joi.array().items(Joi.object({
    name: Joi.string().required(),
    type: Joi.string().required(),
    label: Joi.string().required(),
    description: Joi.string().optional(),
    placeholder: Joi.string().optional(),
    required: Joi.boolean().default(true),
    validation: Joi.object().optional(),
    options: Joi.array().optional()
  })).optional(),
  tags: Joi.string().optional(),
  visibility: Joi.string().valid('public', 'private').default('public'),
});

const promptUpdateSchema = Joi.object({
  title: Joi.string().min(1).max(200).optional(),
  short_description: Joi.string().min(10).max(500).optional(),
  long_description: Joi.string().max(5000).optional(),
  template: Joi.string().min(10).optional(),
  inputs: Joi.array().items(Joi.object({
    name: Joi.string().optional(),
    type: Joi.string().optional(),
    label: Joi.string().optional(),
    description: Joi.string().optional(),
    placeholder: Joi.string().optional(),
    required: Joi.boolean().optional(),
    validation: Joi.object().optional()
  })).optional(),
  tags: Joi.string().optional(),
  visibility: Joi.string().valid('public', 'private').optional(),
});

// Middleware to check if user is creator of prompt
async function checkPromptOwnership(req, res, next) {
  try {
    const prompt = await PromptTemplate.findByPk(req.params.id, {
      include: [{ model: User, as: 'creator' }],
    });
    
    if (!prompt) {
      return res.status(404).json({ error: 'Prompt not found' });
    }
    
    if (prompt.creator_id !== req.user.id) {
      return res.status(403).json({ error: 'Not authorized to modify this prompt' });
    }
    
    req.prompt = prompt;
    next();
  } catch (error) {
    res.status(500).json({ error: 'Server error' });
  }
}

// POST /api/prompts - Create prompt
router.post('/', authenticateToken, async (req, res) => {
  try {
    const { error } = promptCreateSchema.validate(req.body);
    if (error) {
      return res.status(400).json({ error: error.details[0].message });
    }

    const promptData = {
      ...req.body,
      creator_id: req.user.id,
      inputs: req.body.inputs ? JSON.stringify(req.body.inputs) : null,
    };

    const prompt = await PromptTemplate.create(promptData);

    const promptWithCreator = await PromptTemplate.findByPk(prompt.id, {
      include: [{ model: User, as: 'creator', attributes: ['id', 'username'] }],
    });

    res.status(201).json(promptWithCreator);
  } catch (error) {
    console.error('Create prompt error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /api/prompts/favorites - Get user's favorited prompts
router.get('/favorites', authenticateToken, async (req, res) => {
  try {
    const { skip = 0, limit = 20 } = req.query;
    const userId = req.user.id;

    // Get all favorite prompt IDs for the user
    const favorites = await UserPromptFavorites.findAll({
      where: { user_id: userId },
      attributes: ['prompt_id'],
      order: [['prompt_id', 'DESC']],
    });

    const promptIds = favorites.map(fav => fav.prompt_id);
    
    if (promptIds.length === 0) {
      return res.json({
        prompts: [],
        total: 0,
        skip: parseInt(skip),
        limit: parseInt(limit),
      });
    }

    // Get the actual prompts with pagination
    const { count, rows } = await PromptTemplate.findAndCountAll({
      where: {
        id: {
          [Op.in]: promptIds
        }
      },
      include: [
        {
          model: User,
          as: 'creator',
          attributes: ['id', 'username'],
        },
      ],
      order: [['created_at', 'DESC']],
      limit: Math.min(parseInt(limit), 100),
      offset: parseInt(skip),
    });

    res.json({
      prompts: rows,
      total: count,
      skip: parseInt(skip),
      limit: parseInt(limit),
    });
  } catch (error) {
    console.error('Get favorites error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /api/prompts - List prompts with search and filters
router.get('/', authenticateToken, async (req, res) => {
  try {
    const { skip = 0, limit = 20, q, creator_id, tags, favorites_only } = req.query;
    const userId = req.user.id;

    const whereClause = {};
    let joinFavorites = false;

    // If favorites_only filter is requested
    if (favorites_only === 'true') {
      joinFavorites = true;
    }

    // Full-text search across title, short_description, and tags
    if (q) {
      const searchTerm = q.trim();
      console.log('Search term:', searchTerm);
      console.log('Op.or value:', Op.or);
      console.log('Op.and value:', Op.and);
      console.log('Op.iLike value:', Op.iLike);
      
      whereClause[Op.or] = [
        where(fn('UPPER', col('title')), Op.like, `%${searchTerm.toUpperCase()}%`),
        where(fn('UPPER', col('short_description')), Op.like, `%${searchTerm.toUpperCase()}%`),
        where(fn('UPPER', col('tags')), Op.like, `%${searchTerm.toUpperCase()}%`),
      ];
      
      console.log('whereClause keys:', Object.keys(whereClause));
      console.log('whereClause symbols:', Object.getOwnPropertySymbols(whereClause));
      console.log('whereClause[Op.or]:', whereClause[Op.or]);
    }

    // Filter by creator
    if (creator_id) {
      whereClause.creator_id = creator_id;
      console.log('whereClause after creator filter:', JSON.stringify(whereClause, null, 2));
    }

    // Filter by tags (this might conflict with search!)
    if (tags) {
      console.log('Applying tags filter, q exists:', !!q);
      // If we have a search query, we need to be careful not to override the search conditions
      if (q) {
        // Add tags filter to the AND conditions instead
        if (!whereClause[Op.and]) {
          whereClause[Op.and] = [];
        }
        whereClause[Op.and].push(where(fn('UPPER', col('tags')), Op.like, `%${tags.toUpperCase()}%`));
      } else {
        whereClause.tags = where(fn('UPPER', col('tags')), Op.like, `%${tags.toUpperCase()}%`);
      }
      console.log('whereClause after tags filter:', JSON.stringify(whereClause, null, 2));
    }
    
    console.log('whereClause after filters:', JSON.stringify(whereClause, null, 2));

    // Build visibility condition separately and combine with other filters
    const visibilityCondition = {
      [Op.or]: [
        { visibility: 'public' },
        { creator_id: userId },
      ]
    };

    // Combine conditions properly
    let finalWhere = visibilityCondition;
    
    // Check for both string keys and symbol keys (like Op.or)
    const hasConditions = Object.keys(whereClause).length > 0 || Object.getOwnPropertySymbols(whereClause).length > 0;
    
    if (hasConditions) {
      finalWhere = {
        [Op.and]: [
          whereClause,
          visibilityCondition
        ]
      };
    }
    
    console.log('Final where clause has conditions:', hasConditions);
    console.log('Final where symbols:', Object.getOwnPropertySymbols(finalWhere));

    let queryOptions = {
      where: finalWhere,
      include: [
        {
          model: User,
          as: 'creator',
          attributes: ['id', 'username']
        }
      ],
      order: [['created_at', 'DESC']],
      limit: Math.min(parseInt(limit), 100), // Cap limit at 100
      offset: parseInt(skip),
    };

    // If filtering by favorites only, add join with favorites table
    if (joinFavorites) {
      queryOptions.include.push({
        model: User,
        through: UserPromptFavorites,
        as: 'favorited_by',
        where: { id: userId },
        attributes: [],
        required: true
      });
    }

    const { count, rows } = await PromptTemplate.findAndCountAll(queryOptions);

    let promptsWithFavorites;

    if (joinFavorites) {
      // All prompts are favorited since we filtered by favorites
      promptsWithFavorites = rows.map(prompt => {
        const promptData = prompt.toJSON();
        promptData.isFavorited = true;
        return promptData;
      });
    } else {
      // Get user's favorite prompt IDs for status
      const userFavorites = await UserPromptFavorites.findAll({
        where: { user_id: userId },
        attributes: ['prompt_id']
      });
      const favoriteIds = userFavorites.map(fav => fav.prompt_id);

      // Add isFavorited status to each prompt
      promptsWithFavorites = rows.map(prompt => {
        const promptData = prompt.toJSON();
        promptData.isFavorited = favoriteIds.includes(prompt.id);
        return promptData;
      });
    }

    res.json({
      prompts: promptsWithFavorites,
      total: count,
      skip: parseInt(skip),
      limit: parseInt(limit),
    });
  } catch (error) {
    console.error('List prompts error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// GET /api/prompts/:id - Get single prompt
router.get('/:id', authenticateToken, async (req, res) => {
  try {
    const prompt = await PromptTemplate.findByPk(req.params.id, {
      include: [
        { model: User, as: 'creator', attributes: ['id', 'username'] },
      ],
    });

    if (!prompt) {
      return res.status(404).json({ error: 'Prompt not found' });
    }

    // Check visibility
    if (prompt.visibility === 'private' && prompt.creator_id !== req.user.id) {
      return res.status(403).json({ error: 'Private prompt - access denied' });
    }

    // Check if this prompt is favorited by the current user
    const isFavorited = await UserPromptFavorites.findOne({
      where: {
        user_id: req.user.id,
        prompt_id: prompt.id
      }
    });

    const promptData = prompt.toJSON();
    promptData.isFavorited = !!isFavorited;

    res.json(promptData);
  } catch (error) {
    console.error('Get prompt error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// PUT /api/prompts/:id - Update prompt
router.put('/:id', authenticateToken, checkPromptOwnership, async (req, res) => {
  try {
    const { error } = promptUpdateSchema.validate(req.body);
    if (error) {
      return res.status(400).json({ error: error.details[0].message });
    }

    const updateData = {
      ...req.body,
      inputs: req.body.inputs ? JSON.stringify(req.body.inputs) : req.prompt.inputs,
    };

    await req.prompt.update(updateData);

    const updatedPrompt = await PromptTemplate.findByPk(req.prompt.id, {
      include: [{ model: User, as: 'creator', attributes: ['id', 'username'] }],
    });

    res.json(updatedPrompt);
  } catch (error) {
    console.error('Update prompt error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// DELETE /api/prompts/:id - Delete prompt
router.delete('/:id', authenticateToken, checkPromptOwnership, async (req, res) => {
  try {
    await req.prompt.destroy();
    res.status(204).send();
  } catch (error) {
    console.error('Delete prompt error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// POST /api/prompts/:id/favorite - Add to favorites
router.post('/:id/favorite', authenticateToken, async (req, res) => {
  try {
    const promptId = parseInt(req.params.id);
    const userId = req.user.id;

    const prompt = await PromptTemplate.findByPk(promptId);
    if (!prompt) {
      return res.status(404).json({ error: 'Prompt not found' });
    }

    // Check visibility before allowing to favorite
    if (prompt.visibility === 'private' && prompt.creator_id !== userId) {
      return res.status(403).json({ error: 'Private prompt - access denied' });
    }

    // Add to favorites (ignore if already exists)
    const [favorite, created] = await UserPromptFavorites.findOrCreate({
      where: { user_id: userId, prompt_id: promptId }
    });

    if (!created) {
      return res.status(409).json({ error: 'Prompt already in favorites' });
    }

    res.status(201).json({ message: 'Added to favorites' });
  } catch (error) {
    console.error('Add favorite error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

// DELETE /api/prompts/:id/favorite - Remove from favorites
router.delete('/:id/favorite', authenticateToken, async (req, res) => {
  try {
    const promptId = parseInt(req.params.id);
    const userId = req.user.id;

    const result = await UserPromptFavorites.destroy({
      where: {
        user_id: userId,
        prompt_id: promptId,
      },
    });

    if (result === 0) {
      return res.status(404).json({ error: 'Not in favorites' });
    }

    res.status(204).send();
  } catch (error) {
    console.error('Remove favorite error:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});


module.exports = router;